drivers/input/input_internal.c

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright (c) 2022, tangchunhui@coros.com
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "input_internal.h"
+
+int input_internal_setup(struct input_dev *dev)
+{
+	int retval = 0;
+
+	if (dev->buf == NULL) {
+		return -EFAULT;
+	}
+
+	ring_buf_reset(dev->buf);
+
+#ifndef CONFIG_INPUT_ISR_MUTEX
+    retval = k_mutex_init(&(dev->mutex));
+	if (retval < 0) {
+		return retval;
+	}
+#endif
+
+	retval = k_sem_init(&(dev->sem), 0, 1);
+	if (retval < 0) {
+		return retval;
+	}
+
+	dev->config.nonblock = 0;
+
+	return 0;
+}
+
+int input_internal_release(struct input_dev *dev)
+{
+	k_sem_reset(&(dev->sem));
+
+	return 0;
+}
+
+int input_internal_config_get(struct input_dev *dev, struct input_config *config)
+{
+	config->nonblock = dev->config.nonblock;
+
+	return 0;
+}
+
+int input_internal_config_set(struct input_dev *dev, struct input_config *config)
+{
+	dev->config.nonblock = config->nonblock;
+
+	return 0;
+}
+
+#ifdef CONFIG_INPUT_ISR_MUTEX
+
+#define INPUT_LOCK() {		\
+	if (!k_is_in_isr()) {	\
+		key = irq_lock();	\
+	}						\
+}
+
+#define INPUT_UNLOCK() {	\
+	if (!k_is_in_isr()) {	\
+		irq_unlock(key);	\
+	}						\
+}
+
+#else
+
+#define INPUT_LOCK() {									\
+	retval = k_mutex_lock(&(dev->mutex), K_FOREVER);	\
+	if (retval < 0) {									\
+		return retval;									\
+	}													\
+}
+
+#define INPUT_UNLOCK() {			\
+	k_mutex_unlock(&(dev->mutex));	\
+}
+
+#endif
+
+int input_internal_event_read(struct input_dev *dev, struct input_event *event)
+{
+#ifdef CONFIG_INPUT_ISR_MUTEX
+	unsigned int key = 0;
+#endif
+	uint32_t size = 0;
+	int retval = 0;
+
+	if (dev->buf == NULL) {
+		return -EFAULT;
+	}
+
+#ifndef CONFIG_INPUT_ISR_MUTEX
+	INPUT_ASSERT(!k_is_in_isr(), "Can't read event in isr context.");
+#endif
+
+	INPUT_LOCK();
+
+	size = ring_buf_size_get(dev->buf);
+	if (size > 0) {
+		/* Ring buf size must be a multiple of input event size */
+		INPUT_ASSERT((size % input_event_size) == 0, "Invalid ring buf size");
+
+		size = ring_buf_get(dev->buf, (uint8_t *)event, input_event_size);
+	} else if (dev->config.nonblock == false) {
+		do {
+
+			INPUT_UNLOCK();
+
+			retval = k_sem_take(&(dev->sem), K_FOREVER);
+			if (retval < 0) {
+				break;
+			}
+
+			INPUT_LOCK();
+
+			size = ring_buf_size_get(dev->buf);
+			if (size > 0) {
+				/* Ring buf size must be a multiple of input event size */
+				INPUT_ASSERT((size % input_event_size) == 0, "Invalid ring buf size");
+
+				size = ring_buf_get(dev->buf, (uint8_t *)event, input_event_size);
+				break;
+			}
+		} while (1);
+	} else {
+		retval = -EAGAIN;
+	}
+
+	INPUT_UNLOCK();
+
+	return retval;
+}
+
+int input_event(struct input_dev *dev, uint16_t type, uint16_t code, int32_t value)
```

Comment:
creating an event system only for input may be a resource waste.
I would suggest integrating this with system wide event manager. We have the code ready for that (handling input events for both usb and ble) in NCS.